### ISSUE TITLE
[JIT] Decompose Freezing and Constant Propagation Passes

### ISSUE THREAD

**Issue by narendasan** [pytorch/pytorch#37748]
## 🚀 Feature
Separate the constant propagation from the freezing pass.

## Motivation

This will give backends more flexibility in how the graph is lowered and minimizes what the freezing pass does to just freezing attributes. Will allow TRTorch to make sure it lowers to the best performing ops in its lowering pass. Helpful as well for clarity of what each pass does. 

## Pitch

People using the freezing pass will now just expect to get a graph with frozen attributes. People looking to preserve current behavior can then run the constant propagation pass afterwards (hopefully). 

## Alternatives

TRTorch can implement a custom `freeze_module` pass with constant propagation separate or as we did for the time being until the passes are separated just work with the graph given from the pass and take the perf hit on the few cases where TRT Linear out preforms TRT MatMul + Add.  

## Additional context

This is in response to an issue we saw in TRTorch after integrating the `freeze_module` pass, ops like transpose would get optimized out, which in turn masked layers that would normally be picked up by the `FuseLinear` pass, leaving a more general `addmm` operation in its place. While in the long term we will likely move towards lowering to `matmul` -> `add`, right now there are some cases where we may see performance degradation in TensorRT.  

cc @suo

**Comment by SplitInfinity** [comment]
cc: @bzinodev

**Comment by fortianyou** [comment]
I also think it's better this way, separate the constant propagation from the freezing pass. Since the original freezing pass may be error-prone. Here is an example:

```python
import torch

class TestModel(torch.nn.Module):
    
    def forward(self, x):
        return x.size()

x = torch.ones([1, 3, 224, 224])
m = TestModel().eval()
traced_m = torch.jit.trace(m, x)
torch.jit.save(traced_m, 'trace_saved.pt')
saved_m = torch.jit.load('trace_saved.pt')
frozen_traced = torch._C._freeze_module(traced_m._c)
frozen_saved = torch._C._freeze_module(saved_m._c)
print(traced_m.forward.graph)
print(frozen_traced.forward.graph)
print(saved_m.forward.graph)
print(frozen_saved.forward.graph)
````
Outputs:
```
graph(%self : __torch__.___torch_mangle_11.TestModel,
      %x : Float(1:150528, 3:50176, 224:224, 224:1)):
  %3 : int = prim::Constant[value=0]() # <ipython-input-7-54fe5e9e31ae>:6:0
  %4 : int = aten::size(%x, %3) # <ipython-input-7-54fe5e9e31ae>:6:0
  %5 : Long() = prim::NumToTensor(%4)
  %6 : int = prim::Constant[value=1]() # <ipython-input-7-54fe5e9e31ae>:6:0
  %7 : int = aten::size(%x, %6) # <ipython-input-7-54fe5e9e31ae>:6:0
  %8 : Long() = prim::NumToTensor(%7)
  %9 : int = prim::Constant[value=2]() # <ipython-input-7-54fe5e9e31ae>:6:0
  %10 : int = aten::size(%x, %9) # <ipython-input-7-54fe5e9e31ae>:6:0
  %11 : Long() = prim::NumToTensor(%10)
  %12 : int = prim::Constant[value=3]() # <ipython-input-7-54fe5e9e31ae>:6:0
  %13 : int = aten::size(%x, %12) # <ipython-input-7-54fe5e9e31ae>:6:0
  %14 : Long() = prim::NumToTensor(%13)
  %15 : (Long(), Long(), Long(), Long()) = prim::TupleConstruct(%5, %8, %11, %14)
  return (%15)

graph(%self : __torch__.___torch_mangle_13.TestModel,
      %x : Float(1:150528, 3:50176, 224:224, 224:1)):
  %23 : (Long(), Long(), Long(), Long()) = prim::Constant[value=({1}, {3}, {224}, {224})]()
  return (%23)

graph(%self : __torch__.___torch_mangle_11.TestModel,
      %x.1 : Tensor):
  %3 : int = prim::Constant[value=0]() # <ipython-input-7-54fe5e9e31ae>:6:0
  %7 : int = prim::Constant[value=1]() # <ipython-input-7-54fe5e9e31ae>:6:0
  %11 : int = prim::Constant[value=2]() # <ipython-input-7-54fe5e9e31ae>:6:0
  %15 : int = prim::Constant[value=3]() # <ipython-input-7-54fe5e9e31ae>:6:0
  %4 : int = aten::size(%x.1, %3) # <ipython-input-7-54fe5e9e31ae>:6:0
  %5 : Tensor = prim::NumToTensor(%4) # :0:0
  %8 : int = aten::size(%x.1, %7) # <ipython-input-7-54fe5e9e31ae>:6:0
  %9 : Tensor = prim::NumToTensor(%8) # :0:0
  %12 : int = aten::size(%x.1, %11) # <ipython-input-7-54fe5e9e31ae>:6:0
  %13 : Tensor = prim::NumToTensor(%12) # :0:0
  %16 : int = aten::size(%x.1, %15) # <ipython-input-7-54fe5e9e31ae>:6:0
  %17 : Tensor = prim::NumToTensor(%16) # :0:0
  %22 : (Tensor, Tensor, Tensor, Tensor) = prim::TupleConstruct(%5, %9, %13, %17)
  return (%22)

graph(%self : __torch__.___torch_mangle_12.TestModel,
      %x.1 : Tensor):
  %5 : int = prim::Constant[value=3]() # <ipython-input-7-54fe5e9e31ae>:6:0
  %4 : int = prim::Constant[value=2]() # <ipython-input-7-54fe5e9e31ae>:6:0
  %3 : int = prim::Constant[value=1]() # <ipython-input-7-54fe5e9e31ae>:6:0
  %2 : int = prim::Constant[value=0]() # <ipython-input-7-54fe5e9e31ae>:6:0
  %6 : int = aten::size(%x.1, %2) # <ipython-input-7-54fe5e9e31ae>:6:0
  %7 : Tensor = prim::NumToTensor(%6) # :0:0
  %8 : int = aten::size(%x.1, %3) # <ipython-input-7-54fe5e9e31ae>:6:0
  %9 : Tensor = prim::NumToTensor(%8) # :0:0
  %10 : int = aten::size(%x.1, %4) # <ipython-input-7-54fe5e9e31ae>:6:0
  %11 : Tensor = prim::NumToTensor(%10) # :0:0
  %12 : int = aten::size(%x.1, %5) # <ipython-input-7-54fe5e9e31ae>:6:0
  %13 : Tensor = prim::NumToTensor(%12) # :0:0
  %14 : (Tensor, Tensor, Tensor, Tensor) = prim::TupleConstruct(%7, %9, %11, %13)
  return (%14)
```
Like the example above, the original freezing pass seems to do const folding on shapes. But,  the behavior is different if we run freezing pass on a traced TorchScript module which is load from a file. Because the TorchScript's Tensor shapes are eliminated during serialization.

Following is my environments:
```
PyTorch version: 1.6.0a0+9bfb91b
Is debug build: No
CUDA used to build PyTorch: 10.0

OS: Linux Server 7.2 (Paladin)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39)
CMake version: version 3.17.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Tesla T4
GPU 1: Tesla T4

Nvidia driver version: 418.67
cuDNN version: /apsarapangu/cuda-10.0/lib64/libcudnn.so.7.3.1

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] torch==1.6.0a0+9bfb91b
[pip] torchvision==0.7.0a0+11a39aa
[conda] mkl                       2019.0                   pypi_0    pypi
[conda] torch                     1.6.0a0+9bfb91b          pypi_0    pypi
[conda] torchvision               0.7.0a0+11a39aa          pypi_0    pypi
```

**Comment by bzinodev** [comment]
Fixed the issue see https://github.com/pytorch/pytorch/pull/38499
